websockify/websocket.py

```diff
@@ -1,1109 +1,825 @@
-
-        SimpleHTTPRequestHandler.__init__(self, req, addr, server)
-
-    def log_message(self, format, *args):
-        self.logger.info("%s - - [%s] %s" % (self.address_string(), self.log_date_time_string(), format % args))
-
-    @staticmethod
-    def unmask(buf, hlen, plen):
-        pstart = hlen + 4
-        pend = pstart + plen
-        if numpy:
-            b = c = s2b('')
-            if plen >= 4:
-                dtype=numpy.dtype('<u4')
-                if sys.byteorder == 'big':
-                    dtype = dtype.newbyteorder('>')
-                mask = numpy.frombuffer(buf, dtype, offset=hlen, count=1)
-                data = numpy.frombuffer(buf, dtype, offset=pstart,
-                        count=int(plen / 4))
-                #b = numpy.bitwise_xor(data, mask).data
-                b = numpy.bitwise_xor(data, mask).tostring()
+        self._recv_buffer = ''.encode("ascii")
+        self._recv_queue = []
+        self._send_buffer = ''.encode("ascii")
 
-            if plen % 4:
-                #self.msg("Partial unmask")
-                dtype=numpy.dtype('B')
-                if sys.byteorder == 'big':
-                    dtype = dtype.newbyteorder('>')
-                mask = numpy.frombuffer(buf, dtype, offset=hlen,
-                        count=(plen % 4))
-                data = numpy.frombuffer(buf, dtype,
-                        offset=pend - (plen % 4), count=(plen % 4))
-                c = numpy.bitwise_xor(data, mask).tostring()
-            return b + c
-        else:
-            # Slower fallback
-            mask = buf[hlen:hlen+4]
-            data = array.array('B')
-            mask = s2a(mask)
-            data.fromstring(buf[pstart:pend])
-            for i in range(len(data)):
-                data[i] ^= mask[i % 4]
-            return data.tostring()
+        self._sent_close = False
+        self._received_close = False
 
-    @staticmethod
-    def encode_hybi(buf, opcode, base64=False):
-        """ Encode a HyBi style WebSocket frame.
-        Optional opcode:
-            0x0 - continuation
-            0x1 - text frame (base64 encode buf)
-            0x2 - binary frame (use raw buf)
-            0x8 - connection close
-            0x9 - ping
-            0xA - pong
-        """
-        if base64:
-            buf = b64encode(buf)
+        self.close_code = None
+        self.close_reason = None
 
-        b1 = 0x80 | (opcode & 0x0f) # FIN + opcode
-        payload_len = len(buf)
-        if payload_len <= 125:
-            header = pack('>BB', b1, payload_len)
-        elif payload_len > 125 and payload_len < 65536:
-            header = pack('>BBH', b1, 126, payload_len)
-        elif payload_len >= 65536:
-            header = pack('>BBQ', b1, 127, payload_len)
+        self.socket = None
 
-        #self.msg("Encoded: %s", repr(header + buf))
+    def __getattr__(self, name):
+        # These methods are just redirected to the underlying socket
+        if name in ["fileno",
```

Comment:
`SSLSocket` uses internal, undocumented API:s to be able to inherit from `socket`. That may be fine for something part of the Python library, but I don't think we should do that. Not until we're part of the library at least. :)